pkg/agent/monitoraddnodes.go

```diff
@@ -237,3 +265,71 @@ func (mon *addNodeMonitor) isKubeletRunningOnNode() bool {
 	}
 	return false
 }
+
+// decodedFirstCSRSubjectContainsHostname checks if the first CSR
+// is created for the node identified by hostname. It decodes the
+// CSR.Spec.Request PEM block into readable output. When adding a node
+// to a cluster, the first CSR's subject contains the node's hostname.
+//
+// Example of decoded request:
+// Certificate Request:
+// Data:
+// Version: 1 (0x0)
+// Subject: O = system:nodes, CN = system:node:extraworker-1
+// Subject Public Key Info:
+//
+//	Public Key Algorithm: id-ecPublicKey
+//		Public-Key: (256 bit)
+//		pub:
+//			*snip*
+//		ASN1 OID: prime256v1
+//		NIST CURVE: P-256
+//
+// Attributes:
+//
+//	a0:00
+//
+// Signature Algorithm: ecdsa-with-SHA256
+//
+//	*snip*
+func decodedFirstCSRSubjectContainsHostname(request, hostname string) bool {
+	block, _ := pem.Decode([]byte(request))
+	if block == nil {
+		return false
+	}
+	csrDER := block.Bytes
+	decodedRequest, err := x509.ParseCertificateRequest(csrDER)
+	if err != nil {
+		logrus.Warn("error in x509.ParseCertificateRequest(csrDER)")
+		return false
+	}
+	return strings.Contains(decodedRequest.Subject.String(), hostname)
+}
+
+// createIPToHostnameMap generates a map of IP addresses to Hostname
+// using the AddNodesConfig.
+func createIPToHostnameMap(nodesConfig *joiner.AddNodesConfig) (map[string]string, error) {
+	ipToHostname := make(map[string]string)
+
+	for _, host := range nodesConfig.Config.Hosts {
+		if host.Hostname == "" {
+			continue
+		}
+		networkConfigByte, err := yaml.YAMLToJSON(host.NetworkConfig.Raw)
+		if err != nil {
+			return nil, err
+		}
+		networkConfig := string(networkConfigByte)
+
+		result := gjson.Get(networkConfig, `interfaces.#.*.address.#.ip`)
```

Comment:
What if DHCP is used instead?

Reply:
Now that the IP addresses are resolved to hostname, this should take care of the DHCP case.